parsec/serde/schema.py

```diff
@@ -90,26 +92,35 @@ def get_obj_type(self, obj):
     setting `type_field` class property.
     """
 
-    type_field = "type"
-    type_field_remove = True
-    type_schemas = {}
-    fallback_type_schema = None
-    _instantiated_schemas = None
-    _instantiated_fallback_schema = None
+    type_field: str = "type"
+    type_schemas: Dict[Enum, Schema] = {}
+    fallback_type_schema: Schema = None
+    _instantiated_schemas: Dict[Union[Enum, str], Schema] = None
+    _instantiated_fallback_schema: Dict[Union[Enum, str], Schema] = None
+
+    def _build_schemas(self):
+        enum_types = {type(k) for k in self.type_schemas.keys()}
+        if len(enum_types) != 1 or not issubclass(enum_types.pop(), Enum):
+            raise ValueError("type_schemas key can only be one Enum")
+        self._instantiated_schemas = {}
+        for k, v in self.type_schemas.items():
+            schema_instance = v if isinstance(v, Schema) else v()
+            if self.type_field not in schema_instance.fields:
+                raise ValueError(f"{schema_instance} needs to define '{self.type_field}' field")
+            self._instantiated_schemas[k] = schema_instance
+            self._instantiated_schemas[k.value] = schema_instance
 
-    def _get_schema(self, type):
-        if self._instantiated_schemas is None:
-            self._instantiated_schemas = {
-                k: v if isinstance(v, Schema) else v() for k, v in self.type_schemas.items()
-            }
             if self.fallback_type_schema and not isinstance(self.fallback_type_schema, Schema):
```

Comment:
Shouldn't we put this out of the for loop ? (`self._instantiated_fallback_schema` should be instantiated once)
and if we have `self.fallback_type_schema = Myschema()`, then we don't go in this if, hence `self._instantiated_fallback_schema` is `None`...

Reply:
fixed